flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java

```diff
@@ -619,12 +631,17 @@ public void alterPartition(ObjectPath tablePath, CatalogPartitionSpec partitionS
 
   @Override
   public List<String> listFunctions(String dbName) throws CatalogException {
-    return Collections.emptyList();
+    return Lists.newArrayList(partitionFunctions.keySet());
   }
 
   @Override
   public CatalogFunction getFunction(ObjectPath functionPath) throws FunctionNotExistException, CatalogException {
-    throw new FunctionNotExistException(getName(), functionPath);
+    CatalogFunction catalogFunction = partitionFunctions.get(functionPath.getObjectName());
```

Comment:
Are there any situations where this might get used for other reasons than just partition UDFs? Like accessing some other UDF that the user might have registered?

From the existing code, it looks fine. But for my own understanding, I’m still wondering if it’s possible for a user to register their own UDF that would be associated with this Catalog, and if so are we currently making it not possible to do that?

Reply:
> Are there any situations where this might get used for other reasons than just partition UDFs? Like accessing some other UDF that the user might have registered?

Here, we mainly use these udfs to create, modify, and delete partitionSpec through flink.

> From the existing code, it looks fine. But for my own understanding, I’m still wondering if it’s possible for a user to register their own UDF that would be associated with this Catalog, and if so are we currently making it not possible to do that?

Currently, it is not supported for users to register UDF under the flink catalog. But in the next step (just next PR), I will add this feature.

But after we support users to add UDFs, I still want to add these two functions, so that users do not need to manually add them again when they want to query related partitions, I would like to provide them by default. What do you think about this?

Reply:
How Flink users/catalogs handled? Is it possible for one user to access/see other user's functions/catalogs?

Reply:
Reading the doc linked below: https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/functions/overview/, I think I understand it now.

Basically we are registering temporary catalog functions, so the queries will use it anyway, and will not clash with other implementations already registered on the catalog. What happens if the user already registered a `bucket` temporary catalog function?

Why do we use catalog functions instead of system functions? I think the highest priority is the temporary system function, so if we use them we can be reasonably sure that we do not call any user define functions by mistake.

Another issue I see is that if we already has a function registered with the same name, then it might be surprising for the user that the implementation is changed for iceberg tables (not the old catalog level implementation is used, but the one registered by the Iceberg code). Using a more specific name (`iceberg_bucket`), like the one suggested by @kbendick would help to reduce the likelihood of name clashes.

Reply:
I have found the answers for my deleted questions 😄